pipeline-model-definition/src/main/java/org/jenkinsci/plugins/pipeline/modeldefinition/SyntheticStageGraphListener.java

```diff
@@ -54,15 +52,31 @@
     public void onNewHead(FlowNode node) {
         if (node != null && node instanceof StepStartNode &&
                 ((StepStartNode) node).getDescriptor() instanceof StageStep.DescriptorImpl) {
-            LabelAction label = node.getPersistentAction(LabelAction.class);
-            if (label != null &&
-                    (SyntheticStageNames.preStages().contains(label.getDisplayName()) ||
-                            SyntheticStageNames.postStages().contains(label.getDisplayName()))) {
-                if (SyntheticStageNames.preStages().contains(label.getDisplayName())) {
-                    attachTag(node, SyntheticStage.getPre());
+
+            ExecutionModelAction action = null;
+            try {
+                FlowExecutionOwner owner = node.getExecution().getOwner();
+                if (owner != null && owner.getExecutable() instanceof WorkflowRun) {
+                    action = ((WorkflowRun) owner.getExecutable()).getAction(ExecutionModelAction.class);
                 }
-                if (SyntheticStageNames.postStages().contains(label.getDisplayName())) {
-                    attachTag(node, SyntheticStage.getPost());
+            } catch (IOException e) {
+                LOGGER.log(Level.WARNING, "Error loading WorkflowRun for FlowNode: {0}", e);
+            }
+
+            // If the action isn't present, this isn't a Declarative run, so remove the listener.
+            if (action == null) {
+                node.getExecution().removeListener(this);
```

Comment:
Well this definitely looks more brittle than just using `GraphListener` as an `@Extension`.